From f4a9b73e357050882b4a40063edf6fad7c98d543 Mon Sep 17 00:00:00 2001
From: Daniel Ammann <daniel.ammann@bytesatwork.ch>
Date: Tue, 4 Apr 2023 15:16:54 +0200
Subject: [PATCH] byteengine/bytedevkit: Change primary boot medium to eMMC

The SPI-NOR flash will no longer be used from now on. Remove the flash and
the SPI bus.

Switch the environment to eMMC. Use the area between boot sector and first
partition (between 768 KiB and 1 MiB).

Put the bootloader into the eMMC boot0 partition.

Make areas for boot artifacts fit within 4 MiB:
- 512 KiB tiboot3.bin (unchanged*)
- 1.5 MiB tispl.bin (2 MiB)
- 2 MiB u-boot.img (4 MiB)

* The documentation provided by Texas Instruments is partly wrong.

Signed-off-by: Daniel Ammann <daniel.ammann@bytesatwork.ch>
---
 arch/arm/dts/k3-am625-bytedevkit-u-boot.dtsi | 12 ----
 arch/arm/dts/k3-am625-byteengine.dtsi        | 60 +-------------------
 arch/arm/dts/k3-am625-r5-bytedevkit.dts      |  5 --
 board/bytesatwork/bytedevkit/README.md       | 22 +++----
 configs/am62x_bytedevkit_a53_defconfig       | 36 ++----------
 configs/am62x_bytedevkit_r5_defconfig        | 32 +----------
 include/configs/am62x_bytedevkit.h           | 16 +++---
 7 files changed, 30 insertions(+), 153 deletions(-)

diff --git a/arch/arm/dts/k3-am625-bytedevkit-u-boot.dtsi b/arch/arm/dts/k3-am625-bytedevkit-u-boot.dtsi
index 2cbc09faac..92c02e4e3c 100644
--- a/arch/arm/dts/k3-am625-bytedevkit-u-boot.dtsi
+++ b/arch/arm/dts/k3-am625-bytedevkit-u-boot.dtsi
@@ -154,10 +154,6 @@
 	u-boot,dm-spl;
 };
 
-&ospi0_pins_default {
-	u-boot,dm-spl;
-};
-
 &cpsw_port1 {
 	u-boot,dm-spl;
 };
@@ -185,11 +181,3 @@
 		u-boot,dm-spl;
 	};
 };
-
-&ospi0 {
-	u-boot,dm-spl;
-
-	flash@0 {
-		u-boot,dm-spl;
-	};
-};
diff --git a/arch/arm/dts/k3-am625-byteengine.dtsi b/arch/arm/dts/k3-am625-byteengine.dtsi
index 9815235d9d..f88ac6e2d6 100644
--- a/arch/arm/dts/k3-am625-byteengine.dtsi
+++ b/arch/arm/dts/k3-am625-byteengine.dtsi
@@ -11,7 +11,6 @@
 / {
 	aliases {
 		mmc0 = &sdhci0;
-		spi0 = &ospi0;
 	};
 
 	memory@80000000 {
@@ -66,17 +65,6 @@
 			AM62X_IOPAD(0x1f8, PIN_INPUT, 0) /* (AC2) MMC0_DAT7 */
 		>;
 	};
-
-	ospi0_pins_default: ospi0-pins-default {
-		pinctrl-single,pins = <
-			AM62X_IOPAD(0x000, PIN_OUTPUT, 0) /* (H24) OSPI0_CLK */
-			AM62X_IOPAD(0x02c, PIN_OUTPUT, 0) /* (F23) OSPI0_CSn0 */
-			AM62X_IOPAD(0x00c, PIN_INPUT, 0) /* (E25) OSPI0_D0 */
-			AM62X_IOPAD(0x010, PIN_INPUT, 0) /* (G24) OSPI0_D1 */
-			AM62X_IOPAD(0x014, PIN_INPUT, 0) /* (F25) OSPI0_D2 */
-			AM62X_IOPAD(0x018, PIN_INPUT, 0) /* (F24) OSPI0_D3 */
-		>;
-	};
 };
 
 &wkup_uart0 {
@@ -153,51 +141,5 @@
 };
 
 &ospi0 {
-	pinctrl-names = "default";
-	pinctrl-0 = <&ospi0_pins_default>;
-	status = "okay";
-
-	flash@0{
-		compatible = "jedec,spi-nor";
-		reg = <0x0>;
-		spi-tx-bus-width = <4>;
-		spi-rx-bus-width = <4>;
-		spi-max-frequency = <133000000>;
-
-		partitions {
-			compatible = "fixed-partitions";
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			partition@0 {
-				label = "ospi.tiboot3";
-				reg = <0x0 0x80000>;
-			};
-
-			partition@80000 {
-				label = "ospi.tispl";
-				reg = <0x80000 0x200000>;
-			};
-
-			partition@280000 {
-				label = "ospi.u-boot";
-				reg = <0x280000 0x400000>;
-			};
-
-			partition@680000 {
-				label = "ospi.env";
-				reg = <0x680000 0x40000>;
-			};
-
-			partition@6c0000 {
-				label = "ospi.env.backup";
-				reg = <0x6c0000 0x40000>;
-			};
-
-			partition@800000 {
-				label = "ospi.rootfs";
-				reg = <0x800000 0x800000>;
-			};
-		};
-	};
+	status = "disabled";
 };
diff --git a/arch/arm/dts/k3-am625-r5-bytedevkit.dts b/arch/arm/dts/k3-am625-r5-bytedevkit.dts
index e3ada5d50c..4fc7f76968 100644
--- a/arch/arm/dts/k3-am625-r5-bytedevkit.dts
+++ b/arch/arm/dts/k3-am625-r5-bytedevkit.dts
@@ -113,11 +113,6 @@
 	};
 };
 
-&ospi0 {
-	reg = <0x00 0x0fc40000 0x00 0x100>,
-	      <0x00 0x60000000 0x00 0x08000000>;
-};
-
 &main_pktdma {
 	ti,sci = <&dm_tifs>;
 };
diff --git a/board/bytesatwork/bytedevkit/README.md b/board/bytesatwork/bytedevkit/README.md
index 4ab143577f..30a27940f8 100644
--- a/board/bytesatwork/bytedevkit/README.md
+++ b/board/bytesatwork/bytedevkit/README.md
@@ -136,19 +136,21 @@ cp -vp ti-u-boot/build-a53/u-boot.img <path-to-SD-boot-partition>/
 └── ti-u-boot
 ```
 
-## Prepare SPI boot
+## Prepare eMMC boot
 
 Program the `tiboot3.bin`, `tispl.bin` and `u-boot.img` from the SD card to the
-SPI flash.
+eMMC.
 
 ```log
-=> sf probe
-=> fatload mmc 1 ${loadaddr} tiboot3.bin
-=> sf update $loadaddr 0x0 $filesize
-=> fatload mmc 1 ${loadaddr} tispl.bin
-=> sf update $loadaddr 0x80000 $filesize
-=> fatload mmc 1 ${loadaddr} u-boot.img
-=> sf update $loadaddr 0x280000 $filesize
+=> run update_emmc
 ```
 
-**NOTE:** Force SPI boot with the BOOT MODE DIP switches on the byteDEVKIT.
+The bootloader needs to be stored in the boot0 hardware partition of the eMMC.
+The layout of boot0 is defined so that it fits within 4 MiB, defined in blocks
+of 512 Bytes:
+```
+                start   end     size    size
+tiboot3.bin     0x0000  0x0400  0x0400   512 KiB
+tispl.bin       0x0400  0x1000  0x0C00  1536 KiB
+u-boot.img      0x1000  0x2000  0x1000  2048 KiB
+```
diff --git a/configs/am62x_bytedevkit_a53_defconfig b/configs/am62x_bytedevkit_a53_defconfig
index b90e9de8f8..42740c4b40 100644
--- a/configs/am62x_bytedevkit_a53_defconfig
+++ b/configs/am62x_bytedevkit_a53_defconfig
@@ -11,20 +11,15 @@ CONFIG_TARGET_AM62X_BYTEDEVKIT_A53=y
 CONFIG_BAW_COMMON_CMD_OPTIONS=y
 CONFIG_CMD_BAW_CONFIG=y
 CONFIG_ENV_SIZE=0x20000
-CONFIG_ENV_OFFSET=0x680000
-CONFIG_ENV_SECT_SIZE=0x20000
-CONFIG_SYS_SPI_U_BOOT_OFFS=0x280000
-CONFIG_SPL_DM_SPI=y
+CONFIG_ENV_OFFSET=0xC0000
 CONFIG_SPL_TEXT_BASE=0x80080000
 CONFIG_SPL_MMC_SUPPORT=y
 CONFIG_SPL_SERIAL_SUPPORT=y
 CONFIG_SPL_DRIVERS_MISC_SUPPORT=y
 CONFIG_SPL_STACK_R_ADDR=0x82000000
-CONFIG_ENV_OFFSET_REDUND=0x6a0000
+CONFIG_ENV_OFFSET_REDUND=0xE0000
 CONFIG_SPL_FS_FAT=y
 CONFIG_SPL_LIBDISK_SUPPORT=y
-CONFIG_SPL_SPI_FLASH_SUPPORT=y
-CONFIG_SPL_SPI_SUPPORT=y
 CONFIG_DEFAULT_DEVICE_TREE="k3-am625-bytedevkit"
 CONFIG_ENV_VARS_UBOOT_CONFIG=y
 # CONFIG_SYS_MALLOC_CLEAR_ON_INIT is not set
@@ -39,21 +34,17 @@ CONFIG_SPL_SYS_MALLOC_SIMPLE=y
 CONFIG_SPL_STACK_R=y
 CONFIG_SPL_SEPARATE_BSS=y
 CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_USE_SECTOR=y
-CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR=0x1400
+CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR=0x1000
 CONFIG_SPL_DMA=y
 CONFIG_SPL_ENV_SUPPORT=y
 CONFIG_SPL_ETH_SUPPORT=y
 CONFIG_SPL_I2C_SUPPORT=y
 CONFIG_SPL_DM_MAILBOX=y
-CONFIG_SPL_DM_SPI_FLASH=y
 CONFIG_SPL_NET_SUPPORT=y
 CONFIG_SPL_NET_VCI_STRING="AM62X U-Boot A53 SPL"
 CONFIG_SPL_POWER_DOMAIN=y
 CONFIG_SPL_RAM_SUPPORT=y
 CONFIG_SPL_RAM_DEVICE=y
-# CONFIG_SPL_SPI_FLASH_TINY is not set
-CONFIG_SPL_SPI_FLASH_SFDP_SUPPORT=y
-CONFIG_SPL_SPI_LOAD=y
 CONFIG_SPL_THERMAL=y
 CONFIG_SPL_USB_GADGET=y
 CONFIG_SPL_DFU=y
@@ -65,7 +56,6 @@ CONFIG_CMD_ABOOTIMG=y
 CONFIG_CMD_BCB=y
 CONFIG_CMD_DM=y
 # CONFIG_CMD_GPIO is not set
-# CONFIG_CMD_SPI is not set
 CONFIG_CMD_AB_SELECT=y
 CONFIG_CMD_PXE=y
 CONFIG_CMD_SYSBOOT=y
@@ -77,8 +67,9 @@ CONFIG_SPL_OF_CONTROL=y
 CONFIG_MULTI_DTB_FIT=y
 CONFIG_SPL_MULTI_DTB_FIT=y
 CONFIG_SPL_MULTI_DTB_FIT_NO_COMPRESSION=y
-CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_ENV_IS_IN_MMC=y
 CONFIG_SYS_REDUNDAND_ENVIRONMENT=y
+CONFIG_VERSION_VARIABLE=y
 CONFIG_NET_RANDOM_ETHADDR=y
 CONFIG_DM=y
 CONFIG_SPL_DM=y
@@ -93,7 +84,6 @@ CONFIG_SPL_CLK=y
 CONFIG_CLK_TI_SCI=y
 CONFIG_DFU_MMC=y
 CONFIG_DFU_RAM=y
-CONFIG_DFU_SF=y
 CONFIG_SYS_DFU_DATA_BUF_SIZE=0x40000
 CONFIG_SYS_DFU_MAX_FILE_SIZE=0x800000
 CONFIG_DMA_CHANNELS=y
@@ -117,19 +107,6 @@ CONFIG_MMC_SDHCI=y
 CONFIG_MMC_SDHCI_ADMA=y
 CONFIG_SPL_MMC_SDHCI_ADMA=y
 CONFIG_MMC_SDHCI_AM654=y
-CONFIG_DM_SPI_FLASH=y
-CONFIG_SPI_FLASH_SFDP_SUPPORT=y
-CONFIG_SPI_FLASH_SOFT_RESET=y
-CONFIG_SPI_FLASH_SOFT_RESET_ON_BOOT=y
-CONFIG_SPI_FLASH_ATMEL=y
-CONFIG_SPI_FLASH_EON=y
-CONFIG_SPI_FLASH_GIGADEVICE=y
-CONFIG_SPI_FLASH_ISSI=y
-CONFIG_SPI_FLASH_MACRONIX=y
-CONFIG_SPI_FLASH_SPANSION=y
-CONFIG_SPI_FLASH_STMICRO=y
-CONFIG_SPI_FLASH_SST=y
-CONFIG_SPI_FLASH_WINBOND=y
 CONFIG_PHY_MICREL=y
 CONFIG_PHY_MICREL_KSZ8XXX=y
 CONFIG_PHY_TI_DP83867=y
@@ -149,9 +126,6 @@ CONFIG_DM_SERIAL=y
 CONFIG_SOC_DEVICE=y
 CONFIG_SOC_DEVICE_TI_K3=y
 CONFIG_SOC_TI=y
-CONFIG_SPI=y
-CONFIG_DM_SPI=y
-CONFIG_CADENCE_QSPI=y
 CONFIG_SYSRESET=y
 CONFIG_SPL_SYSRESET=y
 CONFIG_SYSRESET_TI_SCI=y
diff --git a/configs/am62x_bytedevkit_r5_defconfig b/configs/am62x_bytedevkit_r5_defconfig
index 855a41ddfd..0548f0cb5e 100644
--- a/configs/am62x_bytedevkit_r5_defconfig
+++ b/configs/am62x_bytedevkit_r5_defconfig
@@ -8,22 +8,17 @@ CONFIG_NR_DRAM_BANKS=2
 CONFIG_SOC_K3_AM625=y
 CONFIG_TARGET_AM62X_BYTEDEVKIT_R5=y
 CONFIG_ENV_SIZE=0x20000
-CONFIG_ENV_OFFSET=0x680000
-CONFIG_ENV_SECT_SIZE=0x20000
-CONFIG_SYS_SPI_U_BOOT_OFFS=0x80000
+CONFIG_ENV_OFFSET=0xC0000
 CONFIG_DM_GPIO=y
-CONFIG_SPL_DM_SPI=y
 CONFIG_SPL_TEXT_BASE=0x43c00000
 CONFIG_SPL_MMC_SUPPORT=y
 CONFIG_SPL_SERIAL_SUPPORT=y
 CONFIG_SPL_DRIVERS_MISC_SUPPORT=y
 CONFIG_SPL_STACK_R_ADDR=0x82000000
 CONFIG_SPL_SIZE_LIMIT=0x38000
-CONFIG_ENV_OFFSET_REDUND=0x6a0000
+CONFIG_ENV_OFFSET_REDUND=0xE0000
 CONFIG_SPL_FS_FAT=y
 CONFIG_SPL_LIBDISK_SUPPORT=y
-CONFIG_SPL_SPI_FLASH_SUPPORT=y
-CONFIG_SPL_SPI_SUPPORT=y
 CONFIG_DEFAULT_DEVICE_TREE="k3-am625-r5-bytedevkit"
 CONFIG_SPL_LOAD_FIT=y
 CONFIG_SPL_LOAD_FIT_ADDRESS=0x80080000
@@ -40,15 +35,11 @@ CONFIG_SPL_DMA=y
 CONFIG_SPL_ENV_SUPPORT=y
 CONFIG_SPL_I2C_SUPPORT=y
 CONFIG_SPL_DM_MAILBOX=y
-CONFIG_SPL_DM_SPI_FLASH=y
 CONFIG_SPL_DM_RESET=y
 CONFIG_SPL_POWER_DOMAIN=y
 CONFIG_SPL_RAM_SUPPORT=y
 CONFIG_SPL_RAM_DEVICE=y
 CONFIG_SPL_REMOTEPROC=y
-# CONFIG_SPL_SPI_FLASH_TINY is not set
-CONFIG_SPL_SPI_FLASH_SFDP_SUPPORT=y
-CONFIG_SPL_SPI_LOAD=y
 CONFIG_SPL_THERMAL=y
 CONFIG_SPL_YMODEM_SUPPORT=y
 CONFIG_HUSH_PARSER=y
@@ -64,7 +55,7 @@ CONFIG_OF_CONTROL=y
 CONFIG_SPL_OF_CONTROL=y
 CONFIG_SPL_MULTI_DTB_FIT=y
 CONFIG_SPL_MULTI_DTB_FIT_NO_COMPRESSION=y
-CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_ENV_IS_IN_MMC=y
 CONFIG_SYS_REDUNDAND_ENVIRONMENT=y
 CONFIG_SYS_RELOC_GD_ENV_ADDR=y
 CONFIG_DM=y
@@ -93,20 +84,6 @@ CONFIG_MMC_SDHCI=y
 CONFIG_MMC_SDHCI_ADMA=y
 CONFIG_SPL_MMC_SDHCI_ADMA=y
 CONFIG_MMC_SDHCI_AM654=y
-CONFIG_DM_SPI_FLASH=y
-CONFIG_SF_DEFAULT_MODE=0
-CONFIG_SPI_FLASH_SFDP_SUPPORT=y
-CONFIG_SPI_FLASH_SOFT_RESET=y
-CONFIG_SPI_FLASH_SOFT_RESET_ON_BOOT=y
-CONFIG_SPI_FLASH_ATMEL=y
-CONFIG_SPI_FLASH_EON=y
-CONFIG_SPI_FLASH_GIGADEVICE=y
-CONFIG_SPI_FLASH_ISSI=y
-CONFIG_SPI_FLASH_MACRONIX=y
-CONFIG_SPI_FLASH_SPANSION=y
-CONFIG_SPI_FLASH_STMICRO=y
-CONFIG_SPI_FLASH_SST=y
-CONFIG_SPI_FLASH_WINBOND=y
 CONFIG_PINCTRL=y
 # CONFIG_PINCTRL_GENERIC is not set
 CONFIG_SPL_PINCTRL=y
@@ -123,9 +100,6 @@ CONFIG_DM_SERIAL=y
 CONFIG_SOC_DEVICE=y
 CONFIG_SOC_DEVICE_TI_K3=y
 CONFIG_SOC_TI=y
-CONFIG_SPI=y
-CONFIG_DM_SPI=y
-CONFIG_CADENCE_QSPI=y
 CONFIG_DM_THERMAL=y
 CONFIG_TIMER=y
 CONFIG_SPL_TIMER=y
diff --git a/include/configs/am62x_bytedevkit.h b/include/configs/am62x_bytedevkit.h
index f7570eac64..1b5a53612b 100644
--- a/include/configs/am62x_bytedevkit.h
+++ b/include/configs/am62x_bytedevkit.h
@@ -88,13 +88,15 @@
 		"load mmc ${mmc_dev}:${mmc_part} ${dtbaddr} ${bootdir}/${dtbfile} || exit; " \
 		"booti ${loadaddr} - ${dtbaddr}; " \
 	"\0" \
-	"update_spi=sf probe; " \
-		"fatload mmc 1 ${loadaddr} tiboot3.bin; " \
-		"sf update $loadaddr 0x0 $filesize; " \
-		"fatload mmc 1 ${loadaddr} tispl.bin; " \
-		"sf update $loadaddr 0x80000 $filesize; " \
-		"fatload mmc 1 ${loadaddr} u-boot.img; " \
-		"sf update $loadaddr 0x280000 $filesize; " \
+	"update_emmc=echo Writing bootloader to eMMC; " \
+		"mmc dev 0 1; " \
+		"load mmc 1 ${loadaddr} tiboot3.bin; " \
+		"mmc write ${loadaddr} 0x0 0x400; " \
+		"load mmc 1 ${loadaddr} tispl.bin; " \
+		"mmc write ${loadaddr} 0x400 0xC00; " \
+		"load mmc 1 ${loadaddr} u-boot.img; " \
+		"mmc write ${loadaddr} 0x1000 0x1000; " \
+		"mmc dev 0 0; " \
 	"\0"
 
 /* The remaining common defines, source <configs/ti_armv7_common.h> */
-- 
2.30.2

